fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -748,11 +783,109 @@ ACTOR Future<Void> commitBatch(
 					}
 					toCommit.addTypedMessage(m);
-				}
-				else
+				} else if (m.type == MutationRef::Exec) {
+					state std::string param2 = m.param2.toString();
+					state ExecCmdValueString execArg(param2);
+					execArg.dbgPrint();
+					state std::string binPath = execArg.getBinaryPath();
+					state std::string uidStr = execArg.getBinaryArgValue("uid");
+
+					if (m.param1 != execDisableTLogPop
+						&& m.param1 != execEnableTLogPop
+						&& !isWhiteListed(self->whiteListedBinPathVec, binPath)) {
+						TraceEvent("ExecTransactionNotPermitted")
+							.detail("TransactionNum", transactionNum);
+						committed[transactionNum] = ConflictBatch::TransactionNotPermitted;
+					} else if (self->db->get().recoveryState != RecoveryState::FULLY_RECOVERED)  {
+						// Cluster is not fully recovered and needs TLogs
+						// from previous generation for full recovery.
+						// Currently, snapshot of old tlog generation is not
+						// supported and hence failing the snapshot request until
+						// cluster is fully_recovered.
+						TraceEvent("ExecTransactionConflict")
+							.detail("TransactionNum", transactionNum);
+						committed[transactionNum] = ConflictBatch::TransactionConflict;
+					} else {
+						// Send the ExecOp to
+						// - all the storage nodes in a single region and
+						// - only to storage nodes in local region in multi-region setup
+						// step 1: get the DatabaseConfiguration
+						state DatabaseConfiguration conf;
+						Standalone<VectorRef<KeyValueRef>> results = wait( self->txnStateStore->readRange( configKeys ) );
+						conf.fromKeyValues(results);
+						// step 2: find the tag.id from locality info of the master
+						auto localityKey =
+							self->txnStateStore->readValue(tagLocalityListKeyFor(self->master.locality.dcId())).get();
+						int8_t locality = tagLocalityInvalid;
+						if (localityKey.present()) {
+							locality = decodeTagLocalityListValue(localityKey.get());
+						}
+
+						auto ranges = self->keyInfo.intersectingRanges(allKeys);
+						std::set<Tag> allSources;
+						auto& m = (*pMutations)[mutationNum];
+						if (debugMutation("ProxyCommit", commitVersion, m))
+							TraceEvent("ProxyCommitTo", self->dbgid)
+								.detail("To", "all sources")
+								.detail("Mutation", m.toString())
+								.detail("Version", commitVersion);
+
+						for (auto r : ranges) {
+							auto& tags = r.value().tags;
+							if (!tags.size()) {
+								for (auto info : r.value().src_info) {
+									if (info->tag.locality == locality) {
+										tags.push_back(info->tag);
+									}
+								}
+								for (auto info : r.value().dest_info) {
+									if (info->tag.locality == locality) {
+										tags.push_back(info->tag);
+									}
+								}
+								uniquify(tags);
+							}
+							std::vector<Tag> localTags;
+							for (auto t : tags) {
+								if ( (!conf.isValid())
+									|| (conf.usableRegions > 1 && t.locality == locality)
+									|| (conf.usableRegions == 1) ) {
```

Comment:
I think if needs to be inside, because I am testing the locality of the tag with the locality of the master. But, the reason I put in the conf.isValid() is because, in one simulator run I got the conf to be invalid.  can conf.isValid() be false in any case?

Reply:
Ah, my eyes appear to have missed that you did have a check against `t`.  `conf.isValid()` can't be false, because otherwise we wouldn't have a valid database here, but I'd probably assert that and throw it through simulation to be sure.